packages/query-core/src/timeoutManager.ts

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * Backend for timer functions.
+ */
+export type TimeoutProvider = {
+  /** Used in error messages. */
+  readonly name: string
+
+  readonly setTimeout: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearTimeout: (timeoutId: number | undefined) => void
+
+  readonly setInterval: (
+    callback: TimeoutCallback,
+    delay: number,
+  ) => TimeoutProviderId
+  readonly clearInterval: (intervalId: number | undefined) => void
+}
+
+export const defaultTimeoutProvider: TimeoutProvider = {
+  name: 'default',
+
+  setTimeout: (callback, delay) => setTimeout(callback, delay),
+  clearTimeout: (timeoutId) => clearTimeout(timeoutId),
+
+  setInterval: (callback, delay) => setInterval(callback, delay),
+  clearInterval: (intervalId) => clearInterval(intervalId),
+}
+
+/** Timeout ID returned by {@link TimeoutManager} */
+export type ManagedTimerId = number
+
+/**
+ * Allows customization of how timeouts are created.
+ *
+ * @tanstack/query-core makes liberal use of timeouts to implement `staleTime`
+ * and `gcTime`. The default TimeoutManager provider uses the platform's global
+ * `setTimeout` implementation, which is known to have scalability issues with
+ * thousands of timeouts on the event loop.
+ *
+ * If you hit this limitation, consider providing a custom TimeoutProvider that
+ * coalesces timeouts.
+ */
+export class TimeoutManager implements Omit<TimeoutProvider, 'name'> {
+  #provider: TimeoutProvider = defaultTimeoutProvider
+  #providerCalled = false
+
+  setTimeoutProvider(provider: TimeoutProvider): void {
+    if (provider === this.#provider) {
+      return
+    }
+
+    if (this.#providerCalled) {
+      // After changing providers, `clearTimeout` will not work as expected for
+      // timeouts from the previous provider.
+      //
+      // Since they may allocate the same timeout ID, clearTimeout may cancel an
+      // arbitrary different timeout, or unexpected no-op.
+      //
+      // We could protect against this by mixing the timeout ID bits
+      // deterministically with some per-provider bits.
+      //
+      // We could internally queue `setTimeout` calls to `TimeoutManager` until
+      // some API call to set the initial provider.
+      console.warn(
+        `[timeoutManager]: Switching to ${provider.name} provider after calls to ${this.#provider.name} provider might result in unexpected behavior.`,
+      )
+    }
+
+    this.#provider = provider
+    this.#providerCalled = false
+  }
+
+  setTimeout(callback: TimeoutCallback, delay: number): ManagedTimerId {
+    this.#providerCalled = true
+    return providerIdToNumber(
+      this.#provider,
+      this.#provider.setTimeout(callback, delay),
+    )
+  }
+
+  clearTimeout(timeoutId: ManagedTimerId | undefined): void {
+    this.#provider.clearTimeout(timeoutId)
+  }
+
+  setInterval(callback: TimeoutCallback, delay: number): ManagedTimerId {
+    this.#providerCalled = true
```

Comment:
those assignments could also be put behind env checks if we move the console error behind the check, too.

Reply:
Done